lib/modules.nix

```diff
@@ -1424,20 +1424,118 @@ let
-    wrap: option:
+    wrapper:
+    mkAliasDefinitionsWith {
+      inherit wrapper;
+      # Set `withLocation = false` to preserve existing behaviour. TODO: Should it be true?
+      withLocation = false;
+    };
+
+  /**
+    A function which transfers all definitions from one option to another.
+    This is useful for renaming options, and also for including properties from
+    another module system, including sub-modules.
+
+    Naturally, definitions are only transferred if the option is defined.
+    Supplying a non-option value is also silently ignored, which is useful for
+    referencing options which may not exist.
+
+    :::{.note}
+
+    This is different than taking the value of the option and using it
+    as a definition, as the new definition will not keep the `mkOverride` /
+    `mkDefault` properties of the previous option.
+
+    If the option you are copying does not have nested properties, or you need
+    to perform some transformation to the value, you may prefer `mkDerivedConfig`,
+    which preserves the option's _overall_ override priority.
+
+    :::
+
+    :::{.caution}
+
+    Transferring definitions to an option with a different `type` may result in
+    the unexpected merge semantics.
+
+    :::
+
+    # Inputs
+
+    `config` (AttrSet)
+
+    : `wrapper` (AttrSet -> Any)
+
+      : A function to wrap the final `mkMerge` definition
+        (e.g. `setAttrByPath [ "foo" "bar" ]`) (default `lib.id`)
+
+      `withPriority` (Bool)
+
+      : Whether to wrap definition with the option's overall priority (`true` by default)
+
+      `withLocation` (Bool)
+
+      : Whether to preserve definitions original file location (`true` by default)
+
+    `option` (Option or Any)
+
+    : The option from which to transfer definitions, if defined.
+
+    # Type
+
+    ```
+    mkAliasDefinitionsWith :: { wrapper :: Function, withPriority :: Bool, withLocation :: Bool } -> Option -> AttrSet
+    ```
+
+    # Example
+
+    ```nix
+    { options, ... }:
+    {
+      # 'bar' might not always be declared in the current module-set.
+      config.foo.enable = mkAliasDefinitionsWith { } (options.bar.enable or { });
+
+      # 'barbaz' has to be declared in the current module-set.
+      config.foobar.paths = mkAliasDefinitionsWith { } options.barbaz.paths;
+    }
+    ```
+  */
+  mkAliasDefinitionsWith =
+    {
+      wrapper ? id,
+      withPriority ? true,
+      withLocation ? true,
+    }:
+    option:
     let
       prio = option.highestPrio or defaultOverridePriority;
-      defsWithPrio = map (mkOverride prio) option.definitions;
+      wrapPrio = if withPriority then mkOverride prio else id;
+      defs =
+        if withLocation then
+          map (def: mkDefinition (def // { value = wrapPrio def.value; })) option.definitionsWithLocations
+        else
+          map wrapPrio option.definitions;
     in
-    mkAliasIfDef option (wrap (mkMerge defsWithPrio));
+    mkAliasIfDef option (wrapper (mkMerge defs));
```

Comment:
Wrapping seems incompatible with `keepPriority`, as the priority is applied to the individual definitions inside, whereas the `foo` and `bar` attributes do not get this treatment.
Maybe the `wrapPrio` should be applied after `wrapper`, and then `pushDownProperties` will(?) take care of the individual pieces.

Reply:
On reflection, I don't think we can drop `wrap` from this function or move it to a private/internal function without making the interface & implementation very clunky.

Since this is already a semi-internal function, would it be acceptable to remove `wrap` from the function-docs and add a comment explaining it exists for internal/compatibility reasons?